src/kitovu/sync/filecache.py

```diff
@@ -0,0 +1,124 @@
+
+8 cases are possible:
+
+Special Cases:
+---------------
+1. remote deleted (triggers exception), local exists: REMOTE_CHANGED
+
+2. remote deleted (triggers exception), local exists AND changed (local_digest and cached_digest differ): BOTH_CHANGED
+case 1 and 2 special, and are dealt with separately in syncing.py.
+
+Cases 1 and 2 are not handled at the moment - see https://jira.keltec.ch/jira/browse/EPJ-77
+
+Normal Cases:
+--------------
+3. new file remote, local none: NEW, download file. This is the default case.
+
+4. remote B, local A - remote digest and local digest differ,
+but local digest and cached digest are same: REMOTE_CHANGED, download.
+
+5. remote and local same: NO_CHANGES
+
+Files changed in-between:
+-------------------------
+6. remote A unchanged, local changed A' - remote digest and cached digest are the same,
+but local digest and cached digest differ: LOCAL_CHANGED
+
+7. remote B, local changed A' - remote digest and cached digest differ,
+local digest and cached digest differ: BOTH_CHANGED
+"""
+
+import attr
+from enum import Enum
+import json
+import pathlib
+import typing
+
+from kitovu.sync import syncplugin
+
+
+class FileState(Enum):
+    """Used to discern in which places files have changed."""
+
+    NEW = 3
+    REMOTE_CHANGED = 4
+    NO_CHANGES = 5
+    LOCAL_CHANGED = 6
+    BOTH_CHANGED = 7
+
+
+@attr.s
+class File:
+
+    cached_digest: str = attr.ib()  # local digest at synctime
+    plugin_name: str = attr.ib()
+
+    def to_dict(self) -> typing.Dict[str, str]:
+        return {"plugin": self.plugin_name,
+                "digest": self.cached_digest}
+
+
+class FileCache:
+
+    def __init__(self, filename: pathlib.Path):
+        self._filename: pathlib.Path = filename
+        self._data: typing.Dict[pathlib.Path, File] = {}
+
+    def _compare_digests(self, remote_digest: str, local_digest: str, cached_digest: str) -> FileState:
+        local_changed: bool = local_digest != cached_digest
+        remote_changed: bool = remote_digest != cached_digest
+        if not remote_changed and not local_changed:  # case 5 above
+            return FileState.NO_CHANGES
+        elif remote_changed and not local_changed:  # case 4 above
+            return FileState.REMOTE_CHANGED
+        elif not remote_changed and local_changed:  # case 6 above
+            return FileState.LOCAL_CHANGED
+        elif remote_changed and local_changed:  # case 7 above
+            return FileState.BOTH_CHANGED
+
+    def write(self) -> None:
+        """"Writes the data-dict to JSON."""
+        json_data: typing.Dict[str, typing.Dict[str, str]] = {}
+
+        for key, value in self._data.items():
+            json_data[str(key)] = value.to_dict()
+
+        with self._filename.open("w") as f:
+            json.dump(json_data, f)
+
+    def load(self) -> None:
+        """This is called first when the synchronisation process is started."""
+        with self._filename.open("r") as f:
```

Comment:
This currently crashes when the filecache doesn't exist initially.

Reply:
fixed in syncing.py